### PR TITLE
Keep track of sent/received/relayed payments and relay/network fees

### PR DESCRIPTION
**Pull request by pm47**
This is a simpler alternative to #505 and is inspired by it.

It introduces a new `AuditDb` which keeps tracks of:
- every single payment (received/sent/relayed)
- every single network fee paid to the miners (funding, closing, and all commit/htlc transactions)

Note that network fees are considered paid when the corresponding tx has reached `min_depth`, it makes sense and allows us to compute the fee in one single place in the `CLOSING` handler. There is an exception for the funding tx, for which we consider the fee paid when the tx has successfully been published to the network. It simplifies the implementation and the tradeoff seems acceptable.

Three new functions have been added to the json-rpc api:
- `audit`: returns all individual payments, with optional filtering on timestamp.
- `networkfees`: returns every single fee paid to the miners, by type (`funding`, `mutual`, `revoked-commit`, etc.) and by channel, with optional filtering on timestamp.
- `channelstats`: maybe the most useful method; it returns a number of information per channel, including the `relayFee` (earned) and the `networkFee` (paid).

The `channels` method now returns details information about channels. It makes it far easier to compute aggregate information about channels using the command line.

Also added a new `ChannelFailed` event that allows e.g. the mobile app to know why a channel got closed.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala [hunk 1]
```diff
@@ -350,6 +361,7 @@ trait Service extends Logging {
     "channels: list existing local channels",
     "channels (nodeId): list existing local channels to a particular nodeId",
     "channel (channelId): retrieve detailed information about a given channel",
+    "channelstats: retrieves statistics about channel usage (fees, number and average amount of payments)",
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: `audit` and `networkfees` are not documented, is it on purpose ?

File: eclair-core/src/main/scala/fr/acinq/eclair/blockchain/electrum/ElectrumEclairWallet.scala [hunk 2]
```diff
@@ -38,7 +38,7 @@ class ElectrumEclairWallet(val wallet: ActorRef, chainHash: BinaryData)(implicit
   override def makeFundingTx(pubkeyScript: BinaryData, amount: Satoshi, feeRatePerKw: Long) = {
     val tx = Transaction(version = 2, txIn = Nil, txOut = TxOut(amount, pubkeyScript) :: Nil, lockTime = 0)
     (wallet ? CompleteTransaction(tx, feeRatePerKw)).mapTo[CompleteTransactionResponse].map(response => response match {
-      case CompleteTransactionResponse(tx1, None) => MakeFundingTxResponse(tx1, 0)
+      case CompleteTransactionResponse(tx1, None) => MakeFundingTxResponse(tx1, 0, Satoshi(tx1.weight() * feeRatePerKw / 1000)) // TODO: output index is always 0? and check fee formula
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/blockchain/electrum/ElectrumEclairWallet.scala]: nitpick: using Transactions.weight2fee() would be cleaner, but modifying ElectrumWallet to return the fee would be even better

File: eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala [hunk 3]
```diff
@@ -833,6 +833,58 @@ object Helpers {
       irrevocablySpent.contains(outPoint)
     }
 
+    /**
+      * This helper function returns the fee paid
+      *
+      * @return
+      */
+    def networkFeePaid(tx: Transaction, d: DATA_CLOSING): Option[(Satoshi, String)] = {
```
Comment (sstone) [inline eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala]: nitpick: scaladoc comment is quite minimalistic (though what the code does is pretty clear)

File: eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala (hunk 1 above)
Comment (pm47) [reply eclair-core/src/main/scala/fr/acinq/eclair/api/Service.scala]: it was, but I changed my mind and they are now documented

**Pull request merged**